loopflow-computation/src/main/java/com/farao_community/farao/loopflow_computation/LoopFlowComputation.java

```diff
@@ -90,13 +94,37 @@ private ReferenceProgramArea gslkToReferenceProgramArea(LinearGlsk glsk) {
         // Extract from the referenceExchangeDataList the ones that are described in the virtualhubs
         List<ReferenceExchangeData> referenceExchangesFromVirtualHubs = referenceProgram.getReferenceExchangeDataList().stream().filter(ReferenceExchangeData::involvesVirtualHub).collect(Collectors.toList());
-        referenceExchangesFromVirtualHubs.forEach(referenceExchangeData -> virtualHubGlsks.add(getVirtualHubGlsk(network, referenceExchangeData)));
-        return new ArrayList<>();
-        // return virtualHubGlsks;
+        List<DanglingLine> danglingLinesWithVirtualHubs = network.getDanglingLineStream().filter(danglingLine -> danglingLine.getExtension(AssignedVirtualHub.class) != null).collect(Collectors.toList());
+        List<Generator> generatorsWithVirtualHubs = network.getGeneratorStream().filter(generator -> generator.getExtension(AssignedVirtualHub.class) != null).collect(Collectors.toList());
+        danglingLinesWithVirtualHubs.forEach(danglingLine -> handleVirtualHubGlsk(virtualHubGlsks, referenceExchangesFromVirtualHubs, danglingLine));
+        generatorsWithVirtualHubs.forEach(generator -> handleVirtualHubGlsk(virtualHubGlsks, referenceExchangesFromVirtualHubs, generator));
+        return virtualHubGlsks;
+    }
+
+    private void handleVirtualHubGlsk(List<LinearGlsk> virtualHubGlsks, List<ReferenceExchangeData> referenceExchangesFromVirtualHubs, Injection injection) {
+        LinearGlsk virtualHubGlsk = getVirtualHubGlsk(injection, referenceExchangesFromVirtualHubs);
+        try {
+            virtualHubGlsks.add(virtualHubGlsk);
+        } catch (FaraoException e) {
+            LOGGER.warn(e.getMessage());
+        }
+    }
+
+    private boolean matchVirtualHub(AssignedVirtualHub assignedVirtualHub, ReferenceExchangeData referenceExchangeData) {
+        return referenceExchangeData.getAreaIn().getAreaCode().equals(assignedVirtualHub.getEic()) || referenceExchangeData.getAreaOut().getAreaCode().equals(assignedVirtualHub.getEic());
     }
 
-    private LinearGlsk getVirtualHubGlsk(Network network, ReferenceExchangeData referenceExchangeData) {
-        return null; //TODO: implement this functio
+    private <T extends Injection> LinearGlsk getVirtualHubGlsk(T virtualHub, List<ReferenceExchangeData> referenceExchangesFromVirtualHubs) {
+        AssignedVirtualHub assignedVirtualHub = (AssignedVirtualHub) virtualHub.getExtension(AssignedVirtualHub.class);
+        List<ReferenceExchangeData> referenceExchanges = referenceExchangesFromVirtualHubs.stream().filter(referenceExchangeData -> matchVirtualHub(assignedVirtualHub, referenceExchangeData)).collect(Collectors.toList());
+        if (referenceExchanges.size() != 1) {
+            String exceptionMessage = "Unable to create a virtual hub for generator " + virtualHub.getId();
+            throw new FaraoException(exceptionMessage);
+        } else {
+            Map<String, Float> glskMap = new HashMap<>();
+            glskMap.put(virtualHub.getId(), 1.0F);
```

Comment:
The id of the GLSK map should be one of a generator or a load in the network.
Here, our sensitivityAnalysis will not be able to map the `virtualHub.getId()` on the network.

You could look here : https://github.com/farao-community/farao-core/blob/master/data/glsk/glsk-document-api/src/main/java/com/farao_community/farao/data/glsk/api/util/converters/GlskPointLinearGlskConverter.java

lines 95 to 102 -> We need to look in the network a generator which is located in the virtual Hub, and put the id of this generator in the glskMap.

If the AssignedVirtualHub is on a generator -> it's not gonna so difficult
If the AssignedVirtualHub is on a danglingLine -> it's gonna be more tricky